src/rgen/Microsoft.Macios.Bindings.Analyzer/NativeObjectHandleAnalyzer.cs

```diff
@@ -0,0 +1,166 @@
+
+	public override void Initialize (AnalysisContext context)
+	{
+		context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.None);
+		context.EnableConcurrentExecution ();
+		context.RegisterSyntaxNodeAction (AnalyzeNode, SyntaxKind.SimpleMemberAccessExpression);
+	}
+
+	private bool IsINativeObject (ITypeSymbol typeSymbol)
+	{
+		if (typeSymbol.TypeKind == TypeKind.TypeParameter)
+			return ((ITypeParameterSymbol) typeSymbol).ConstraintTypes.Any (t => IsINativeObject (t));
+		if (typeSymbol.TypeKind == TypeKind.Interface && typeSymbol.ToDisplayString () == "ObjCRuntime.INativeObject")
+			return true;
+		return typeSymbol.AllInterfaces.Any (i => i.ToDisplayString () == "ObjCRuntime.INativeObject");
+	}
+
+	private bool IsHandleAccessor (MemberAccessExpressionSyntax memberAccess)
+	{
+		string name = memberAccess.Name.Identifier.Text;
+		switch (name) {
+		case "Handle":
+		case "GetHandle":
+		case "GetNonNullHandle":
+		case "GetCheckedHandle":
+			return true;
+		default:
+			return false;
+		}
+	}
+
+	private bool IsSafeMethod (IMethodSymbol methodSymbol)
+	{
+		var name = methodSymbol.Name;
+		// White-list couple of methods that return one of the input parameters or a static
+		// object.
+		switch (name) {
+		case "ThrowOnNull":
+		case "DangerousAutorelease":
+		case "GetConstant":
+			return true;
+		default:
+			return false;
+		}
+	}
+
+	private ITypeSymbol GetRealReturnType (IMethodSymbol methodSymbol)
+	{
+		if (methodSymbol.MethodKind == MethodKind.Constructor)
+			return methodSymbol.ContainingType;
+		return methodSymbol.ReturnType;
+	}
+
+	private void AnalyzeNode (SyntaxNodeAnalysisContext context)
+	{
+		if (context.Node is not MemberAccessExpressionSyntax memberAccess)
+			return;
+
+		if (memberAccess.Name.Identifier.Text != "Handle" &&
+			memberAccess.Name.Identifier.Text != "GetHandle" &&
+			memberAccess.Name.Identifier.Text != "GetNonNullHandle" &&
+			memberAccess.Name.Identifier.Text != "GetCheckedHandle")
+			return;
+
+		var symbol = context.SemanticModel.GetSymbolInfo (memberAccess.Expression).Symbol;
+		if (symbol is null)
+			return;
+
+		if ((symbol as IParameterSymbol)?.IsThis == true)
+			return;
+
+		if (symbol is IMethodSymbol methodSymbol) {
+			if (IsINativeObject (GetRealReturnType (methodSymbol)) &&
+				!IsSafeMethod (methodSymbol)) {
+				// Calling Handle directly on a value returned from method is wrong because
+				// the object is immediately collectible.                
+				var diagnostic = Diagnostic.Create (RBI0014, memberAccess.GetLocation (), symbol.Name);
+				context.ReportDiagnostic (diagnostic);
+			}
+			return;
+		}
+
+		var type = (symbol as ILocalSymbol)?.Type ?? (symbol as IParameterSymbol)?.Type;
+		if (type is null || !IsINativeObject (type))
+			return;
+
+		// Ignore variables that are wrapped in using block
+		var usingStatement = memberAccess.Parent?.FirstAncestorOrSelf<UsingStatementSyntax> ();
```

Comment:
Mentioned in the original PR, we need to check if the declaration syntax has a using keyword too.

Reply:
I think we can just check `ILocalSymbol.IsUsing` instead. I'll try it.

Reply:
Yes, it's implemented and covered by the unit tests.